### PR TITLE
Add ability to restart a session on a fatal error

### PR DESCRIPTION
**Pull request by robskillington**

### CODE REVIEW

File: Jetstream/Client.swift [hunk 1]
```diff
@@ -72,16 +75,22 @@ public enum ClientStatus {
...
     
     // MARK: - Public interface
     
     /// Constructs a client.
     ///
     /// :param: transportAdapter The transport adapter to use to connect to the Jetstream server.
-    public init(transportAdapter: TransportAdapter) {
-        transport = Transport(adapter: transportAdapter)
+    /// :param: restartSessionOnFatalError Whether to restart a session on a fatal session or transport error.
+    public init(transportAdapterFactory: TransportAdapterFactory, restartSessionOnFatalError: Bool = true) {
```
Comment (artman) [inline Jetstream/Client.swift]: Should the transportAdapterFactory be the last argument as is usually the case with closures?

Comment (robskillington) [inline Jetstream/Client.swift]: The second parameter is optional though, hm.

File: Jetstream/Session.swift [hunk 2]
```diff
@@ -33,14 +33,13 @@ public class Session {
     let client: Client
     var nextMessageIndex: UInt = 1
     var serverIndex: UInt = 0
-    var scopes = [UInt: Scope]()
+    var scopes = [UInt: (Scope, [String: AnyObject])]()
```
Comment (artman) [inline Jetstream/Session.swift]: You should give names to the variables in the tuple.

File: Jetstream/Session.swift [hunk 3]
```diff
@@ -108,7 +107,7 @@ public class Session {
         
         switch message {
         case let scopeStateMessage as ScopeStateMessage:
-            if let scope = scopes[scopeStateMessage.scopeIndex] {
+            if let scope = scopes[scopeStateMessage.scopeIndex]?.0 {
```
Comment (artman) [inline Jetstream/Session.swift]: If your tuple variables have name you don't need to use 0, but can use its name.

File: Jetstream/Session.swift [hunk 4]
```diff
@@ -120,7 +119,7 @@ public class Session {
                 logger.error("Received state message without having local scope")
             }
         case let scopeSyncMessage as ScopeSyncMessage:
-            if let scope = scopes[scopeSyncMessage.scopeIndex] {
+            if let scope = scopes[scopeSyncMessage.scopeIndex]?.0 {
```
Comment (artman) [inline Jetstream/Session.swift]: Here, too

File: Jetstream/Session.swift [hunk 5]
```diff
@@ -164,7 +163,7 @@ public class Session {
     
     func close() {
         for (_, scope) in scopes {
-            scope.onChanges.removeListener(self)
+            scope.0.onChanges.removeListener(self)
```
Comment (artman) [inline Jetstream/Session.swift]: And here

File: Jetstream/Client.swift (hunk 1 above)
Comment (artman) [inline Jetstream/Client.swift]: Yeah, no worries then

File: Jetstream/Client.swift [hunk 6]
```diff
@@ -72,16 +75,22 @@ public enum ClientStatus {
     }
     
     let logger = Logging.loggerFor("Client")
-    let transport: Transport
+    let transportAdapterFactory: TransportAdapterFactory
+    let restartSessionOnFatalError: Bool
+    var transport: Transport
     var sessionCreateParams = [String: AnyObject]()
+    var explicitlyClosed = false
```
Comment (artman) [inline Jetstream/Client.swift]: This isn't used anywhere

File: JetstreamTests/TestTransportAdapter.swift [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+import Foundation
+import Signals
```
Comment (artman) [inline JetstreamTests/TestTransportAdapter.swift]: Signals are directly imported to the project, so you shouldn't import. Don't know why it doesn't create a compilation error on your side

File: JetstreamTests/ClientTests.swift [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+import Foundation
+import UIKit
+import XCTest
+import Signals
```
Comment (artman) [inline JetstreamTests/ClientTests.swift]: Signals are directly imported to the project, so you shouldn't import. Don't know why it doesn't create a compilation error on your side

**Comment by artman** [comment]
Done with first round

File: Jetstream/Session.swift (hunk 2 above)
Comment (robskillington) [inline Jetstream/Session.swift]: Sadly this is not working, I've tried a multitude of different ways even using Typealiases.  It keeps saying "Expected ',' separator.  Invalid character in source file":

```swift
typealias ScopeWithFetchParams = (scope: Scope, fetchParams: [String: AnyObject])
// ..
    var scopes = [UInt: ScopeWithFetchParams]()
```

File: Jetstream/Client.swift (hunk 6 above)
Comment (robskillington) [inline Jetstream/Client.swift]: Good catch, removing now

File: JetstreamTests/TestTransportAdapter.swift (hunk 7 above)
Comment (robskillington) [inline JetstreamTests/TestTransportAdapter.swift]: Will remove

File: JetstreamTests/ClientTests.swift (hunk 8 above)
Comment (robskillington) [inline JetstreamTests/ClientTests.swift]: Will remove

File: Jetstream/Session.swift (hunk 2 above)
Comment (robskillington) [inline Jetstream/Session.swift]: Never mind fixed, weird source character creeped into my file that was invisible

**Comment by robskillington** [comment]
Updated but found one of the other tests crashing, will investigate.  Wonder why Travis CI is not running =/

**Comment by artman** [comment]
Looks good once test passes.

**Pull request merged**